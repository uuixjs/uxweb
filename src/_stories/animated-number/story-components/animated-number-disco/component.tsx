import * as React from "react";

import { AnimatedNumber, Layout } from "v2";

import { Component } from "react";

interface DiscoProps {
  startValue: number;
  deltaMin?: number;
  deltaMax?: number;
}

interface DiscoState {
  value: number;
}

export class AnimatedNumberDisco extends Component<DiscoProps, DiscoState> {
  private interval: number | undefined;

  public constructor(props: DiscoProps) {
    super(props);
    this.state = {
      value: props.startValue,
    };
  }

  public componentDidMount() {
    this.interval = window.setInterval(this.tick, 2000);
  }

  public componentWillUnmount() {
    window.clearInterval(this.interval);
  }

  public render() {
    return (
      <Layout>
        <AnimatedNumber
          value={this.state.value}
          formatValue={(n) => n.toLocaleString()}
        />
      </Layout>
    );
  }

  private tick = () => {
    const diff = generateRandomInteger(
      this.props.deltaMin || -10,
      this.props.deltaMax || 10,
    );

    this.setState((prevState) => ({
      value: prevState.value + diff,
    }));
  };
}

function generateRandomInteger(min: number, max: number) {
  return Math.floor(min + Math.random() * (max + 1 - min));
}
