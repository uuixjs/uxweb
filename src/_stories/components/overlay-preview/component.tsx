import "./styles.scss";

import * as React from "react";

import {
  AlignItems,
  Background,
  CoreImage,
  Display,
  InjectLayout,
  Layout,
  Overflow,
  OverlayRegion,
  Position,
  Radio,
  TransitionGroup,
  TransitionType,
  ZIndex,
} from "v2";
import { Component, FormEvent } from "react";
import { Padding, cn, newUUIDv4 } from "@uuixjs/uuixweb-lib";

import { ExampleThumbnails } from "../../assets";

export enum ExampleBG {
  Off = "non-overlay",
  Photo1 = "photo-1",
  Photo2 = "photo-2",
  Video = "video",
  Black = "black",
  White = "white",
}

interface Props {
  defaultBg?: ExampleBG;
  blur?: boolean;
  padding?: Padding;
}

interface State {
  bg: ExampleBG;
}

export class OverlayPreview extends Component<Props, State> {
  public static defaultProps = {
    padding: 2,
    blur: true,
  };

  private uid = newUUIDv4();

  public constructor(props: Props) {
    super(props);
    this.state = {
      bg: props.defaultBg || ExampleBG.Black,
    };
  }

  public render() {
    const overlay = this.state.bg !== ExampleBG.Off;

    return (
      <Layout>
        <Layout display={Display.Flex} padding={1}>
          {Object.keys(ExampleBG).map((k, i) => (
            <Layout key={k} padding={{ left: i > 0 ? 1 : 0 }}>
              <Radio
                name={`overlay-preview-${this.uid}`}
                label={k}
                value={ExampleBG[k as keyof typeof ExampleBG]}
                onChange={this.onRadioSelected}
                checked={
                  this.state.bg === ExampleBG[k as keyof typeof ExampleBG]
                }
              />
            </Layout>
          ))}
        </Layout>

        <OverlayRegion overlay={overlay}>
          <Layout
            className="overlay-preview__container"
            position={Position.Relative}
            background={overlay ? Background.Overlay : undefined}
            padding={this.props.padding}
          >
            {this.renderBackground()}
            <Layout zIndex={ZIndex.Above} position={Position.Relative}>
              {this.props.children}
            </Layout>
          </Layout>
        </OverlayRegion>
      </Layout>
    );
  }

  private renderBackground() {
    let element;
    if (this.state.bg === ExampleBG.Photo1) {
      element = <CoreImage src={ExampleThumbnails.stream1} alt="" />;
    }

    if (this.state.bg === ExampleBG.Photo2) {
      element = <CoreImage src={ExampleThumbnails.stream2} alt="" />;
    }

    if (this.state.bg === ExampleBG.Video) {
      element = (
        <video
          src="https://d3hmvhl7ru3t12.cloudfront.net/video/pages/home/header-82b9abce067992beedde6d103b6241d616a7ee97cc0d4f3126277098c2e4b09004325f25c6a245656096bd333c880b0032e2cbdbed9d2514220e20718482e95b.mp4"
          autoPlay
          muted
          loop
        />
      );
    }

    if (this.state.bg === ExampleBG.Black) {
      element = <div style={{ background: "#000" }} />;
    }

    if (this.state.bg === ExampleBG.White) {
      element = <div style={{ background: "#FFF" }} />;
    }

    return (
      <InjectLayout
        display={Display.Flex}
        alignItems={AlignItems.Stretch}
        attachTop
        attachLeft
        attachBottom
        attachRight
        overflow={Overflow.Hidden}
      >
        <TransitionGroup
          className="overlay-preview__transition-group-container"
          transitionType={TransitionType.Fade}
        >
          {element ? (
            <InjectLayout
              className={cn({
                ["overlay-preview__bg-element"]: true,
                ["overlay-preview__bg-element--blur"]: this.props.blur,
              })}
              key={this.state.bg}
              position={Position.Absolute}
              attachTop
              attachLeft
              attachBottom
              attachRight
              fullWidth
              fullHeight
            >
              {element}
            </InjectLayout>
          ) : (
            []
          )}
        </TransitionGroup>
      </InjectLayout>
    );
  }

  private onRadioSelected = (e: FormEvent<HTMLInputElement>) => {
    if (e.currentTarget.checked) {
      this.setState({
        bg: e.currentTarget.value as ExampleBG,
      });
    }
  };
}
