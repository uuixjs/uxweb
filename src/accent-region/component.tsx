import {
  AccentToken,
  AccentTokenValues,
  accentTokenDarkThemeMap,
  accentTokenLightThemeMap,
  formatAccentTokenCssVars,
  formatTokenOverrides,
} from "./token-map";
import { RGBColor, parse, rgb2hex } from "lib/pure-color";
import {
  ThemeContextValue,
  accentColorGenerator,
  colorIsDark,
  formatColor,
  getHoverColor,
  styled,
  useTheme,
} from "@uuixjs/uuixweb-lib";

import { FC } from "react";
import { StaticTokenMap } from "lib/ui-tokens";
import { TokenOverrideRegion } from "../token-override-region";

export interface AccentRegionProps {
  /** The original chosen color */
  inputColor?: string | null;
  /** Set to true if the contrast ratio between the original color and `#FFFFFF` is less than 4.5 */
  inputColorIsDark?: boolean | null;
  /** Color generated according to hover color spec */
  inputColorHover?: string | null;
  /** Color generated according to `accentColorGenerator` spec */
  primary1?: string | null;
  /** Color generated according to `accentColorGenerator` spec */
  primary2?: string | null;
  /** Color generated according to `accentColorGenerator` spec */
  primary3?: string | null;
  /** Color generated according to `accentColorGenerator` spec */
  primary4?: string | null;
  /** Color generated according to `accentColorGenerator` spec */
  primary5?: string | null;
}

export const ScAccentRegionCssVars = styled.div<{ $values: AccentTokenValues }>`
  /** Accent Token CSS Vars: */
  ${({ $values }) => formatAccentTokenCssVars($values)}
`;

/**
 * Changes the color palette for a region of the page
 * and should be populated with values which have been generated
 * according to the Core UI accent color generator spec.
 *
 * Values may be generated and stored on the server, or can be generated
 * at render time by using the functions provided in the Core UI package.
 */
export const AccentRegion: FC<AccentRegionProps> = ({
  inputColor,
  inputColorIsDark,
  inputColorHover,
  primary1,
  primary2,
  primary3,
  primary4,
  primary5,
  children,
}) => {
  const tokenValues: AccentTokenValues = {};

  if (inputColor) {
    tokenValues[AccentToken.Accent] = formatColor(inputColor);
  }

  if (typeof inputColorIsDark === "boolean") {
    tokenValues[AccentToken.AccentLabel] = inputColorIsDark
      ? "#FFFFFF"
      : "#000000";
  }

  if (inputColorHover) {
    tokenValues[AccentToken.Hover] = formatColor(inputColorHover);
  }

  if (typeof inputColorIsDark === "boolean") {
    tokenValues[AccentToken.BackgroundButtonPurchase] = inputColorIsDark
      ? StaticTokenMap["color-opac-b-5"]
      : StaticTokenMap["color-opac-w-5"];
  }

  if (primary1) {
    tokenValues[AccentToken.Primary1] = formatColor(primary1);
  }

  if (primary2) {
    tokenValues[AccentToken.Primary2] = formatColor(primary2);
  }

  if (primary3) {
    tokenValues[AccentToken.Primary3] = formatColor(primary3);
  }

  if (primary4) {
    tokenValues[AccentToken.Primary4] = formatColor(primary4);
  }

  if (primary5) {
    tokenValues[AccentToken.Primary5] = formatColor(primary5);
  }

  const theme = useTheme();
  let tokenOverrides: ThemeContextValue["tokenOverrides"];

  if (theme === undefined || theme.name === "dark") {
    tokenOverrides = formatTokenOverrides(tokenValues, accentTokenDarkThemeMap);
  } else if (theme.name === "light") {
    tokenOverrides = formatTokenOverrides(
      tokenValues,
      accentTokenLightThemeMap,
    );
  }

  return (
    <TokenOverrideRegion tokenOverrides={tokenOverrides}>
      <ScAccentRegionCssVars $values={tokenValues}>
        {children}
      </ScAccentRegionCssVars>
    </TokenOverrideRegion>
  );
};

export function generateAccentRegionProps(
  color: string | null,
): AccentRegionProps {
  const rgb = color && (parse(formatColor(color)) as RGBColor);

  if (!rgb) {
    return {
      inputColor: null,
      inputColorIsDark: false,
      inputColorHover: null,
      primary1: null,
      primary2: null,
      primary3: null,
      primary4: null,
      primary5: null,
    };
  }

  const colors = accentColorGenerator(rgb).map((v) => rgb2hex(v));

  return {
    inputColor: rgb2hex(rgb),
    inputColorIsDark: colorIsDark(rgb),
    inputColorHover: rgb2hex(getHoverColor(rgb)),
    primary1: colors[0],
    primary2: colors[1],
    primary3: colors[2],
    primary4: colors[3],
    primary5: colors[4],
  };
}
