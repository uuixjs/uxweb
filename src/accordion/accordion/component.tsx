import { AccordionBody, AccordionBodyProps } from "../accordion-body";
import { AccordionFooter, AccordionFooterProps } from "../accordion-footer";
import { AccordionHeader, AccordionHeaderProps } from "../accordion-header";
import { BorderRadius, getAriaProps } from "@uuixjs/uuixweb-lib";
import { BorderRadiusProps, Layout } from "../../layout";
import { Component, MouseEvent } from "react";

export type AccordionItem = {
  header: AccordionHeaderProps;
  body: AccordionBodyProps;
  footer?: AccordionFooterProps;
};

export interface AccordionProps {
  items: AccordionItem[];
  initialOpenIndex?: number | false;
}

export interface AccordionState {
  currentOpenIndex: number | false;
}

/**
 * Accordion is used to consistently compose and control some accordion items
 * which will each be composed of an AccordionHeader, AccordionBody, and AccordionFooter
 */
export class Accordion extends Component<AccordionProps, AccordionState> {
  public state: AccordionState = {
    currentOpenIndex:
      this.props.initialOpenIndex !== undefined
        ? this.props.initialOpenIndex
        : 0,
  };

  public render() {
    return (
      <Layout {...getAriaProps(this.props)}>
        {this.props.items.map((item, index) => (
          <Layout
            borderBottom={index !== this.props.items.length - 1}
            key={index}
          >
            <AccordionHeader
              {...item.header}
              isOpen={index === this.state.currentOpenIndex}
              onClick={this.onHeaderClick}
              index={index}
              borderRadius={this.getHeaderBorderRadius(index)}
            />

            <AccordionBody
              {...item.body}
              isOpen={index === this.state.currentOpenIndex}
              borderRadius={this.getBodyBorderRadius(index)}
            >
              {item.body.children}
              {item.footer && <AccordionFooter {...item.footer} />}
            </AccordionBody>
          </Layout>
        ))}
      </Layout>
    );
  }

  private onHeaderClick = (_EVENT: MouseEvent<HTMLElement>, index?: number) => {
    if (typeof index === "number") {
      this.setState((prevState) => ({
        currentOpenIndex: prevState.currentOpenIndex === index ? false : index,
      }));
    }
  };

  private getHeaderBorderRadius = (
    index: number,
  ): BorderRadiusProps | undefined => {
    const corners: BorderRadiusProps = {};

    if (index === 0) {
      corners.topLeft = BorderRadius.Medium;
      corners.topRight = BorderRadius.Medium;
    }

    if (
      index === this.props.items.length - 1 &&
      this.state.currentOpenIndex !== index
    ) {
      corners.bottomLeft = BorderRadius.Medium;
      corners.bottomRight = BorderRadius.Medium;
    }

    return corners;
  };

  private getBodyBorderRadius = (
    index: number,
  ): BorderRadiusProps | undefined => {
    if (index === this.props.items.length - 1) {
      return {
        bottomLeft: BorderRadius.Medium,
        bottomRight: BorderRadius.Medium,
      };
    }

    return undefined;
  };
}
