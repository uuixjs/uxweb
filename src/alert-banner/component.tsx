import {
  AlignItems,
  Color,
  Display,
  FlexDirection,
  FontSize,
  InjectLayout,
  JustifyContent,
  Layout,
  Position,
} from "../layout";
import { Button, ButtonPublicProps, ButtonType } from "../button/button";
import { CloseButtonProps, CoreDismissible } from "../core-dismissible";
import { CoreLink, CoreLinkType } from "../core-link";
import { CoreText, TextType } from "../core-text";
import { StatusButton, StatusButtonProps } from "../button/status-button";
import { styleVariant, styled, themeTokenRule } from "@uuixjs/uuixweb-lib";

import { Column } from "../grid/column";
import { CoreInteractivePublicProps } from "../core-interactive";
import { FC } from "react";
import { Grid } from "../grid/grid";
import { Icon } from "../icon";
import { NotificationType } from "../notifications/notification";
import { ResponsiveWrapper } from "../responsive-wrapper";
import { SVGAsset } from "../svg";

export type AlertBannerActions =
  | [Omit<StatusButtonProps, "type">]
  | [Omit<StatusButtonProps, "type">, ButtonPublicProps];

export interface AlertBannerLinkProps extends CoreInteractivePublicProps {
  text: string;
}

export interface AlertBannerProps {
  /**
   * Up to two actionable buttons as part of the banner.
   */
  actions?: AlertBannerActions | JSX.Element;
  /**
   * Renders a close button target.
   */
  closeButton?: CloseButtonProps;
  /**
   * Optional link CTA for further context.
   */
  link?: AlertBannerLinkProps;
  /**
   * A brief description of state of the alert taking place.
   *
   * @example The System is Down.
   */
  status?: string;
  /**
   * Alert message text.
   *
   * @example Please stand by, we will inform you when it is back up.
   */
  message: string | JSX.Element;
  /**
   * The type of alert
   *
   * @example Error
   */
  type: AlertBannerType | "error" | "info" | "success" | "warning"; // TODO: https://jira.xarth.tv/browse/COREUI-3554
}

export enum AlertBannerType {
  Error = "error",
  Info = "info",
  Success = "success",
  Warning = "warning",
}

const ALERT_BANNER_ICON: Record<AlertBannerType, SVGAsset> = {
  [AlertBannerType.Error]: SVGAsset.NotificationError,
  [AlertBannerType.Info]: SVGAsset.NotificationInfo,
  [AlertBannerType.Success]: SVGAsset.NotificationSuccess,
  [AlertBannerType.Warning]: SVGAsset.NotificationWarning,
};

const NOTIFICATION_TYPE_MAP: Record<AlertBannerType, NotificationType> = {
  [AlertBannerType.Error]: NotificationType.Error,
  [AlertBannerType.Info]: NotificationType.Info,
  [AlertBannerType.Success]: NotificationType.Success,
  [AlertBannerType.Warning]: NotificationType.Warning,
};

const ScAlertBanner = styled(Layout)<{ $type: AlertBannerProps["type"] }>`
  background-color: ${themeTokenRule("color-background-base")};
  border-bottom-width: 5px;
  border-bottom-style: solid;
  border-bottom-color: ${styleVariant("$type", {
    [AlertBannerType.Error]: themeTokenRule("color-border-error-decorative"),
    [AlertBannerType.Info]: themeTokenRule("color-border-info-decorative"),
    [AlertBannerType.Success]: themeTokenRule(
      "color-border-success-decorative",
    ),
    [AlertBannerType.Warning]: themeTokenRule("color-border-warn-decorative"),
  })};
`;

const ScAlertBannerIcon = styled.div<{ $type: AlertBannerProps["type"] }>`
  display: flex;
  align-items: center;
  flex-shrink: 0;
  color: ${styleVariant("$type", {
    [AlertBannerType.Error]: themeTokenRule("color-fill-error"),
    [AlertBannerType.Info]: themeTokenRule("color-fill-info"),
    [AlertBannerType.Success]: themeTokenRule("color-fill-success"),
    [AlertBannerType.Warning]: themeTokenRule("color-fill-warn"),
  })};
`;

export const AlertBanner: FC<AlertBannerProps> = ({
  link,
  closeButton,
  type,
  actions,
  status,
  message,
  ...props
}) => {
  let linkElement: JSX.Element | undefined;
  const actionsElements: JSX.Element[] = [];

  if (link) {
    linkElement = (
      <InjectLayout margin={{ left: 1 }}>
        <CoreLink
          className="tw-alert-banner__link"
          variant={CoreLinkType.Inherit}
          underline
          hoverColorInherit
          {...link}
        >
          {link.text}
        </CoreLink>
      </InjectLayout>
    );
  }

  if (actions && Array.isArray(actions)) {
    actions.forEach((action, i) => {
      actionsElements.push(
        <Layout key={i} margin={{ left: 1 }}>
          {i === 0 ? (
            <StatusButton {...action} type={NOTIFICATION_TYPE_MAP[type]} />
          ) : (
            <Button {...action} variant={ButtonType.Secondary} />
          )}
        </Layout>,
      );
    });
  } else if (actions) { // @ts-ignore
    actionsElements.push(actions);
  }

  return (
    <ScAlertBanner
      {...props}
      $type={type}
      role="alert"
      className="tw-alert-banner"
      fontSize={FontSize.Size6}
      alignItems={AlignItems.Center}
      position={Position.Relative}
      display={Display.Flex}
      fullWidth
      padding={{ y: 0.5, left: closeButton && 0.5 }}
    >
      <ResponsiveWrapper centered>
        <Grid justifyContent={JustifyContent.Center}>
          <Column cols={{ default: 12, lg: 9, xl: 8, xxl: 7 }}>
            <Layout
              display={Display.Flex}
              alignItems={AlignItems.Center}
              position={Position.Relative}
            >
              <ScAlertBannerIcon $type={type}>
                <Icon asset={ALERT_BANNER_ICON[type]} />
              </ScAlertBannerIcon>
              <Layout
                display={Display.Flex}
                alignItems={AlignItems.Center}
                flexGrow={1}
              >
                {status && (
                  <InjectLayout margin={{ left: 1 }}>
                    <CoreText type={TextType.H2} bold fontSize={FontSize.Size6}>
                      {status}
                    </CoreText>
                  </InjectLayout>
                )}
                {typeof message !== "string" && (
                  <>
                    {message}
                    {linkElement}
                  </>
                )}
                {message && typeof message === "string" && (
                  <Layout margin={{ left: 1 }}>
                    <CoreText
                      type={TextType.P}
                      fontSize={FontSize.Size6}
                      color={Color.Alt2}
                    >
                      {message}
                      {linkElement}
                    </CoreText>
                  </Layout>
                )}
              </Layout>
              {actionsElements && (
                <Layout
                  display={Display.Flex}
                  flexDirection={FlexDirection.RowReverse}
                  justifyContent={JustifyContent.End}
                >
                  {actionsElements}
                </Layout>
              )}
              {closeButton && (
                <Layout padding={{ left: 1 }}>
                  <CoreDismissible
                    aria-label={closeButton["aria-label"]}
                    onClick={closeButton.onClick}
                  />
                </Layout>
              )}
            </Layout>
          </Column>
        </Grid>
      </ResponsiveWrapper>
    </ScAlertBanner>
  );
};

AlertBanner.displayName = "AlertBanner";
