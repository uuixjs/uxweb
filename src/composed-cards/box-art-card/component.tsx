import { Aspect, AspectRatio } from "../../aspect";
import {
  BorderRadius,
  getAriaProps,
  getDataProps,
  hoverCss,
  styled,
  themeTokenRule,
} from "lib";
import { Color, Display, FontSize, Layout } from "../../layout";
import { CoreLink, CoreLinkType } from "../../core-link";
import { CoreText, LineHeight, TextType } from "../../core-text";
import { Tag, TagProps } from "../../tag";

import { Card } from "../../card/card";
import { CardBody } from "../../card/card-body";
import { CardImage } from "../../card/card-image";
import { CoreImageProps } from "../../core-image";
import { CoreInteractivePublicProps } from "../../core-interactive";
import { FC } from "react";
import { Placeholder } from "../../placeholder";
import { getImageProps } from "../../../_utils/image-props";

export interface BoxArtCardProps
  extends CoreImageProps,
    Omit<CoreInteractivePublicProps, "refHandler"> {
  /**
   * The meta information displayed under the title.
   *
   * @example 106,303 viewers
   */
  info: string | JSX.Element;
  /** Show the placeholder representation of this card */
  placeholder?: boolean;
  /** A list of tag interfaces to render into tags underneath the card body */
  tags?: TagProps[];
  /**
   * The card's title.
   *
   * @example League of Legends
   */
  title: string;
  /** Displays the BoxArt as unavailable to launch */
  unavailable?: boolean;
}

const ScBoxArtCardImage = styled.div<{
  unavailable?: BoxArtCardProps["unavailable"];
  disabled?: BoxArtCardProps["disabled"];
}>`
  ${({ unavailable, disabled }) =>
    unavailable &&
    !disabled && {
      opacity: 0.5,
    }}
`;

const ScBoxArtCardLink = styled(CoreLink)`
  ${hoverCss`
    color: ${themeTokenRule("color-text-base")};
  `}

  ${({ disabled }) =>
    disabled && {
      cursor: "not-allowed",
      opacity: 0.5,
    }};
`;

/**
 * Box art cards can be used to link to game and category views.
 *
 * @deprecated Please use the new components named MediaCard, MediaCardImage, MediaCardMeta, etc.
 */
export const BoxArtCard: FC<BoxArtCardProps> = ({
  unavailable,
  disabled,
  ...props
}) => {
  const cardContent = (
    <>
      <ScBoxArtCardImage
        className="tw-box-art-card__image"
        data-test-selector="tw-card-image"
        data-a-target="tw-card-image"
        unavailable={unavailable}
        disabled={disabled}
      >
        <CardImage
          aspect={AspectRatio.BoxArt}
          borderRadius={BorderRadius.Medium}
          {...getImageProps(props)}
        />
      </ScBoxArtCardImage>
      <CardBody>
        <Layout
          data-test-selector="tw-card-title"
          data-a-target="tw-card-title"
          margin={{ top: 0.5 }}
        >
          <CoreText
            className="tw-box-art-card__title"
            type={TextType.H3}
            fontSize={FontSize.Size5}
            lineHeight={LineHeight.Body}
            ellipsis
          >
            {props.title}
          </CoreText>
          <CoreText color={Color.Alt2} ellipsis>
            {props.info}
          </CoreText>
        </Layout>
      </CardBody>
    </>
  );

  if (props.placeholder) {
    return (
      <>
        <Layout margin={{ bottom: 0.5 }}>
          <Aspect ratio={AspectRatio.Aspect3x4}>
            <Placeholder />
          </Aspect>
        </Layout>
        <CoreText>
          <Placeholder width={150} />
        </CoreText>
        <CoreText fontSize={FontSize.Size7}>
          <Placeholder width={100} />
        </CoreText>
      </>
    );
  }

  return (
    <div
      className="tw-box-art-card"
      {...getDataProps(props)}
      {...getAriaProps(props)}
    >
      <Card key={props.title}>
        {props.linkTo || props.onClick || disabled ? (
          <ScBoxArtCardLink
            className="tw-box-art-card__link"
            data-a-target="tw-box-art-card-link"
            disabled={disabled}
            download={props.download}
            linkTo={props.linkTo}
            aria-label={props.title}
            onClick={props.onClick}
            renderLink={props.renderLink}
            targetBlank={props.targetBlank}
            tabIndex={props.tabIndex}
            variant={CoreLinkType.Inherit}
            hoverColorInherit={disabled || unavailable}
            hoverUnderlineNone
          >
            {cardContent}
          </ScBoxArtCardLink>
        ) : (
          cardContent
        )}
        {props.tags && (
          <Layout className="tw-box-art-card__tags" margin={{ top: 0.5 }}>
            {props.tags.map((tag, i) => (
              <Layout
                key={i}
                fontSize={FontSize.Size7}
                display={Display.InlineBlock}
                margin={{ right: 0.5, bottom: 0.5 }}
              >
                <Tag {...tag} data-a-target={tag.label} />
              </Layout>
            ))}
          </Layout>
        )}
      </Card>
    </div>
  );
};

BoxArtCard.displayName = "BoxArtCard";
