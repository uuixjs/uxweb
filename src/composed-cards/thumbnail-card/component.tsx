import { Aspect, AspectRatio } from "../../aspect";
import { BorderRadius, getAriaProps, getDataProps } from "lib";
import { Color, FontSize, Layout } from "../../layout";
import { CoreLink, CoreLinkProps, CoreLinkType } from "../../core-link";
import { CoreText, LineHeight, TextType } from "../../core-text";

import { Card } from "../../card/card";
import { CardBody } from "../../card/card-body";
import { CardImage } from "../../card/card-image";
import { CoreImageProps } from "../../core-image";
import { FC } from "react";
import { Placeholder } from "../../placeholder";
import { getImageProps } from "../../_utils/image-props";

export interface ThumbnailCardProps
  extends CoreImageProps,
    Pick<CoreLinkProps, "linkTo"> {
  /** The meta information displayed under the title. */
  info: string | JSX.Element;
  /** Show the placeholder representation of this card */
  placeholder?: boolean;
  /** The card's title. */
  title: string;
}

/**
 * @deprecated Please use the new components named MediaCard, MediaCardImage, MediaCardMeta, etc.
 */
export const ThumbnailCard: FC<ThumbnailCardProps> = (props) => {
  const cardContent = (
    <Card key={props.title}>
      <CardImage
        aspect={AspectRatio.Aspect16x9}
        borderRadius={BorderRadius.Medium}
        {...getImageProps(props)}
      />
      <CardBody>
        <Layout margin={{ top: 0.5 }}>
          <CoreText
            className="tw-thumbnail-card__title"
            type={TextType.H3}
            fontSize={FontSize.Size5}
            lineHeight={LineHeight.Body}
            ellipsis
          >
            {props.title}
          </CoreText>
        </Layout>
        <CoreText color={Color.Alt2} ellipsis>
          {props.info}
        </CoreText>
      </CardBody>
    </Card>
  );

  if (props.placeholder) {
    return (
      <Layout>
        <Layout margin={{ bottom: 0.5 }}>
          <Aspect ratio={AspectRatio.Aspect16x9}>
            <Placeholder />
          </Aspect>
        </Layout>
        <CoreText>
          <Placeholder width={150} />
        </CoreText>
        <CoreText fontSize={FontSize.Size7}>
          <Placeholder width={100} />
        </CoreText>
      </Layout>
    );
  }

  if (props.linkTo) {
    return (
      <div
        className="tw-thumbnail-card"
        {...getDataProps(props)}
        {...getAriaProps(props)}
      >
        <CoreLink
          className="tw-thumbnail-card__link"
          linkTo={props.linkTo}
          title={props.title}
          data-a-target="tw-thumbnail-card-link"
          variant={CoreLinkType.Inherit}
          hoverUnderlineNone
        >
          {cardContent}
        </CoreLink>
      </div>
    );
  } else {
    return (
      <div className="tw-thumbnail-card" {...getDataProps(props)}>
        {cardContent}
      </div>
    );
  }
};

ThumbnailCard.displayName = "ThumbnailCard";
