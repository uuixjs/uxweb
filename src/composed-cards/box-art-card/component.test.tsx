import { setupShallowTest } from "../../../tests/helpers";
import { BoxArtCard, BoxArtCardProps } from "./component";

function getRequiredProps(): BoxArtCardProps {
  return {
    info: "test sting",
    title: "Test title",
    src: "srcURL",
    alt: "Test Alt",
  };
}

function getPlaceholderProps(): BoxArtCardProps {
  return {
    info: "test sting",
    placeholder: true,
    title: "Test title",
    src: "srcURL",
    alt: "Test Alt",
  };
}

function getOptionalProps(): BoxArtCardProps {
  return {
    info: "test sting",
    linkTo: "/",
    title: "Test title",
    src: "srcURL",
    alt: "Test Alt",
    disabled: true,
    unavailable: true,
    tags: [
      {
        label: "Tag One",
      },
      {
        label: "Tag Two",
      },
    ],
  };
}

const setupShallow = setupShallowTest(BoxArtCard, getRequiredProps);
describe("BoxArtCard", () => {
  it("renders box art card component with required props", () => {
    const { wrapper } = setupShallow();
    expect(wrapper).toMatchSnapshot();
  });

  it("renders box art card component with all props", () => {
    const { wrapper } = setupShallow(getOptionalProps());
    expect(wrapper).toMatchSnapshot();
  });

  it("renders box art card component as a placeholder", () => {
    const { wrapper } = setupShallow(getPlaceholderProps());
    expect(wrapper).toMatchSnapshot();
  });

  it("has data-a-target on the title", () => {
    const { wrapper } = setupShallow();
    expect(wrapper.find('[data-test-selector="tw-card-title"]')).toHaveProp(
      "data-a-target",
      "tw-card-title",
    );
  });

  it("has a data-a-target on the image", () => {
    const { wrapper } = setupShallow();
    expect(wrapper.find('[data-test-selector="tw-card-image"]')).toHaveProp(
      "data-a-target",
      "tw-card-image",
    );
  });
});
