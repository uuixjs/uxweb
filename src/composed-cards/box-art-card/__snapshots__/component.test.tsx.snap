// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoxArtCard renders box art card component as a placeholder 1`] = `
<Fragment>
  <Layout
    margin={
      Object {
        "bottom": 0.5,
      }
    }
  >
    <Aspect
      align={3}
      overflow={false}
      ratio="tw-aspect--3x4"
    >
      <Placeholder />
    </Aspect>
  </Layout>
  <CoreText>
    <Placeholder
      width={150}
    />
  </CoreText>
  <CoreText
    fontSize="font-size-7"
  >
    <Placeholder
      width={100}
    />
  </CoreText>
</Fragment>
`;

exports[`BoxArtCard renders box art card component with all props 1`] = `
<div
  className="tw-box-art-card"
>
  <Card
    key="Test title"
  >
    <Styled(Component)
      aria-label="Test title"
      className="tw-box-art-card__link"
      data-a-target="tw-box-art-card-link"
      disabled={true}
      hoverColorInherit={true}
      hoverUnderlineNone={true}
      linkTo="/"
      variant="inherit"
    >
      <styled.div
        className="tw-box-art-card__image"
        data-a-target="tw-card-image"
        data-test-selector="tw-card-image"
        disabled={true}
        unavailable={true}
      >
        <CardImage
          alt="Test Alt"
          aspect="tw-aspect--3x4"
          borderRadius="border-radius-medium"
          src="srcURL"
        />
      </styled.div>
      <CardBody>
        <Layout
          data-a-target="tw-card-title"
          data-test-selector="tw-card-title"
          margin={
            Object {
              "top": 0.5,
            }
          }
        >
          <CoreText
            className="tw-box-art-card__title"
            ellipsis={true}
            fontSize="font-size-5"
            lineHeight="body"
            type="h3"
          >
            Test title
          </CoreText>
          <CoreText
            color="alt-2"
            ellipsis={true}
          >
            test sting
          </CoreText>
        </Layout>
      </CardBody>
    </Styled(Component)>
    <Layout
      className="tw-box-art-card__tags"
      margin={
        Object {
          "top": 0.5,
        }
      }
    >
      <Layout
        display="inline-block"
        fontSize="font-size-7"
        key="0"
        margin={
          Object {
            "bottom": 0.5,
            "right": 0.5,
          }
        }
      >
        <Tag
          data-a-target="Tag One"
          label="Tag One"
        />
      </Layout>
      <Layout
        display="inline-block"
        fontSize="font-size-7"
        key="1"
        margin={
          Object {
            "bottom": 0.5,
            "right": 0.5,
          }
        }
      >
        <Tag
          data-a-target="Tag Two"
          label="Tag Two"
        />
      </Layout>
    </Layout>
  </Card>
</div>
`;

exports[`BoxArtCard renders box art card component with required props 1`] = `
<div
  className="tw-box-art-card"
>
  <Card
    key="Test title"
  >
    <styled.div
      className="tw-box-art-card__image"
      data-a-target="tw-card-image"
      data-test-selector="tw-card-image"
    >
      <CardImage
        alt="Test Alt"
        aspect="tw-aspect--3x4"
        borderRadius="border-radius-medium"
        src="srcURL"
      />
    </styled.div>
    <CardBody>
      <Layout
        data-a-target="tw-card-title"
        data-test-selector="tw-card-title"
        margin={
          Object {
            "top": 0.5,
          }
        }
      >
        <CoreText
          className="tw-box-art-card__title"
          ellipsis={true}
          fontSize="font-size-5"
          lineHeight="body"
          type="h3"
        >
          Test title
        </CoreText>
        <CoreText
          color="alt-2"
          ellipsis={true}
        >
          test sting
        </CoreText>
      </Layout>
    </CardBody>
  </Card>
</div>
`;
